win32/getopt.h

```diff
@@ -0,0 +1,902 @@
+     * external `optind' increments; (`argind' must lag by one).
+     */
+    argind = optind - 1;
+
+    /* When `optind' is misused, in this fashion, we also abandon any
+     * residual text in the argument we had been parsing; this is done
+     * without any further processing of such abandoned text, assuming
+     * that the caller is equipped to handle it appropriately.
+     */
+    nextchar = NULL;
+  }
+
+  if( nextchar && *nextchar )
+  {
+    /* we are parsing a standard, or short format, option argument ...
+     */
+    const CHAR *optchar;
+    if( (optchar = getopt_match( optopt = *nextchar++, optstring )) != NULL )
+    {
+      /* we have identified it as valid ...
+       */
+      if( optchar[1] == getopt_takes_argument )
+      {
+	/* and determined that it requires an associated argument ...
+	 */
+	if( ! *(optarg = (char *)(nextchar)) )
+	{
+	  /* the argument is NOT attached ...
+	   */
+	  if( optchar[2] == getopt_takes_argument )
+	    /*
+	     * but this GNU extension marks it as optional,
+	     * so we don't provide one on this occasion.
+	     */
+	    optarg = NULL;
+
+	  /* otherwise this option takes a mandatory argument,
+	   * so, provided there is one available ...
+	   */
+	  else if( (argc - argind) > 1 )
+	    /*
+	     * we take the following command line argument,
+	     * as the appropriate option argument.
+	     */
+	    optarg = argv[++argind];
+
+	  /* but if no further argument is available,
+	   * then there is nothing we can do, except for
+	   * issuing the requisite diagnostic message.
+	   */
+	  else
+	  {
+	    complain( "option requires an argument -- %c", optopt );
+	    return getopt_missing_arg( optstring );
+	  }
+	}
+	optind = argind + 1;
+	nextchar = NULL;
+      }
+      else
+	optarg = NULL;
+      optind = (nextchar && *nextchar) ? argind : argind + 1;
+      return optopt;
+    }
+    /* if we didn't find a valid match for the specified option character,
+     * then we fall through to here, so take appropriate diagnostic action.
+     */
+    if( mode == getopt_mode_long_only )
+    {
+      complain( "unrecognised option `-%s'", --nextchar );
+      nextchar = NULL;
+      optopt = 0;
+    }
+    else
+      complain( "invalid option -- %c", optopt );
+    optind = (nextchar && *nextchar) ? argind : argind + 1;
+    return getopt_unknown;
+  }
+
+  if( optmark > optbase )
+  {
+    /* This can happen, in GNU parsing mode ONLY, when we have
+     * skipped over non-option arguments, and found a subsequent
+     * option argument; in this case we permute the arguments.
+     */
+    int index;
+    /*
+     * `optspan' specifies the number of contiguous arguments
+     * which are spanned by the current option, and so must be
+     * moved together during permutation.
+     */
+    const int optspan = argind - optmark + 1;
+    /*
+     * we use `this_arg' to store these temporarily.
+     */
+    CHAR *this_arg = malloc(sizeof(CHAR) * optspan);
```

Comment:
I think this should be `CHAR **this_arg = malloc(sizeof(CHAR*) * optspan);` instead, which should fix some warnings